contracts/MarketOracle.sol

```diff
@@ -14,104 +14,128 @@ interface IOracle {
+    external
+    onlyOwner
+    {
+        require(reportExpirationTimeSec_ > 0);
+        reportExpirationTimeSec = reportExpirationTimeSec_;
+    }
+
+    function setMinimumProviders(uint256 minimumProviders_)
+    external
+    onlyOwner
+    {
+        require(minimumProviders_ > 0);
+        minimumProviders = minimumProviders_;
+    }
+
+    function lookupProviderIndex(address provider) internal view returns (uint256) {
+        uint256 position = whitelistMap[provider];
+        require(position != 0);
+        return position.sub(1);
+    }
+
+    function setProviderIndex(address provider, uint256 index) internal {
+        whitelistMap[provider] = index.add(1);
+    }
+
+    function pushReport(uint256 payload) external
+    {
+        uint256 index = lookupProviderIndex(msg.sender);
+        Report storage report = whitelistReports[index];
+        report.timestamp = block.timestamp;
+        report.payload = payload;
+    }
 
-    /**
-     * @dev Calculates the volume weighted average of exchange rates and total trade volume.
-     *      Expired market sources are ignored. If there has been no trade volume in the last
-     *      24hrs, then there is effectively no exchange rate and that value should be ignored by
-     *      the client.
-     * @return exchangeRate: Volume weighted average of exchange rates.
-    *          isValid: True if data is fresh and false if not.
-     */
     function getData()
         external
         returns (uint256, bool)
     {
-        uint256 volumeWeightedSum = 0;
-        uint256 volumeSum = 0;
-        uint256 partialRate = 0;
-        uint256 partialVolume = 0;
-        bool isSourceFresh = false;
-
-        for (uint256 i = 0; i < whitelist.length; i++) {
-            (isSourceFresh, partialRate, partialVolume) = whitelist[i].getReport();
-
-            if (!isSourceFresh) {
-                emit LogSourceExpired(whitelist[i]);
-                continue;
+        uint256[] memory validReports = new uint256[](whitelistReports.length);
+        uint256 size = 0;
+        for (uint256 i = 0; i < whitelistReports.length; i++) {
+            if (now < reportExpirationTimeSec.add(whitelistReports[i].timestamp)) {
+                validReports[size++] = whitelistReports[i].payload;
+            } else {
+                emit LogSourceExpired(whitelistReports[i].provider);
             }
+        }
 
-            volumeWeightedSum = volumeWeightedSum.add(partialRate.mul(partialVolume));
-            volumeSum = volumeSum.add(partialVolume);
+        if (size < minimumProviders) {
+            return (0, false);
         }
 
-        if (volumeSum > 0) {
-            // No explicit fixed point normalization is done as dividing by volumeSum normalizes
-            // to exchangeRate's format.
-            return (volumeWeightedSum.div(volumeSum), true);
+        uint256 median;
+        Select.quickSort(validReports, 0, size - 1);
```

Comment:
This code block is one of the few places we use unchecked math. AFAICT, everything's safe as long as size > 0.

We should probably document our assumptions here in case, for example, the minimumProviders check were to change.

Reply:
Was not planning to commit this code as is. Will rewrite once  Sort/Select function is finalized.